lib/sms/sms_submit.c

```diff
@@ -0,0 +1,354 @@
+ *                     If send_at_cmds is true, this is input.
+ *                     If send_at_cmds is false, this is output.
+ */
+static int sms_submit_concat(
+	const char *text,
+	uint8_t *encoded_number,
+	uint8_t encoded_number_size,
+	uint8_t encoded_number_size_octets,
+	bool send_at_cmds,
+	uint8_t* concat_msg_count)
+{
+	int err = 0;
+	char user_data[SMS_MAX_PAYLOAD_LEN_CHARS];
+	char udh_str[13] = {0};
+
+	static uint8_t concat_msg_id = 1;
+	static uint8_t message_ref = 1;
+	uint8_t concat_seq_number = 0;
+
+	uint8_t size;
+	uint16_t text_size = strlen(text);
+	uint8_t encoded_data_size_octets = 0;
+	uint8_t encoded_data_size_septets = 0;
+
+	uint16_t text_encoded_size = 0;
+	const char *text_index = text;
+	uint16_t text_part_size;
+
+	memset(user_data, 0, SMS_UDH_CONCAT_SIZE_SEPTETS);
+
+	/* More message parts are created until entire input text is encoded into messages */
+	while (text_encoded_size < text_size) {
+		concat_seq_number++;
+
+		text_part_size = MIN(strlen(text_index), SMS_MAX_CONCAT_PAYLOAD_LEN_CHARS);
+		memcpy(user_data + SMS_UDH_CONCAT_SIZE_SEPTETS, text_index, text_part_size);
+
+		/* User Data Header is included into encoded buffer because the actual data/text
+		 * must be aligned into septet (7bit) boundary after User Data Header.
+		 */
+		size = string_conversion_ascii_to_gsm7bit(user_data,
+			SMS_UDH_CONCAT_SIZE_SEPTETS + text_part_size,
+			sms_payload_tmp, &encoded_data_size_octets, &encoded_data_size_septets,
+			true);
+
+		/* User Data Header septets must be ignored from the text size and indexing */
+		text_encoded_size += size - SMS_UDH_CONCAT_SIZE_SEPTETS;
+		text_index += size - SMS_UDH_CONCAT_SIZE_SEPTETS;
+
+		if (send_at_cmds) {
+			sprintf(udh_str, "050003%02X%02X%02X",
+				concat_msg_id, *concat_msg_count, concat_seq_number);
+
+			err = sms_submit_encode(
+				sms_buf_tmp,
+				encoded_number,
+				encoded_number_size,
+				encoded_number_size_octets,
+				sms_payload_tmp + SMS_UDH_CONCAT_SIZE_OCTETS,
+				encoded_data_size_octets - SMS_UDH_CONCAT_SIZE_OCTETS,
+				encoded_data_size_septets,
+				message_ref,
+				udh_str);
+
+			if (err) {
+				break;
+			}
+
+			message_ref++;
+
+			/* Just looping without threading seems to work fine and we don't need to wait
+			* for CDS response. Otherwise we would need to send 2nd message from work queue
+			* and store a lot of state information.
+			*/
+		}
+	}
+
+	if (send_at_cmds) {
+		concat_msg_id++;
+	} else {
+		*concat_msg_count = concat_seq_number;
+	}
+	return err;
+}
+
+int sms_submit_send(const char *number, const char *text)
+{
+	char empty_string[] = "";
+	int err;
+	uint8_t encoded_number[SMS_MAX_ADDRESS_LEN_CHARS + 1];
+	uint8_t encoded_number_size;
+	uint8_t encoded_number_size_octets = SMS_MAX_ADDRESS_LEN_CHARS + 1;
+	uint8_t size = 0;
+	uint16_t text_size;
+	uint8_t encoded_data_size_octets = 0;
+	uint8_t encoded_data_size_septets = 0;
+	uint8_t concat_msg_count = 0;
+
+	if (number == NULL) {
+		number = empty_string;
```

Comment:
It was centrally blocked in sms_submit_encode_number() but I changed it here so that we return instantly with an error so it's more obvious for the reader.